portal/src/hook/useSessionStorage.ts

```diff
@@ -0,0 +1,179 @@
+
+      return parsed as T;
+    },
+    [options, initialValue]
+  );
+
+  // Get from session storage then
+  // parse stored json or return initialValue
+  const readValue = useCallback((): T => {
+    const initialValueToUse =
+      initialValue instanceof Function ? initialValue() : initialValue;
+
+    try {
+      const raw = window.sessionStorage.getItem(key);
+      return raw ? deserializer(raw) : initialValueToUse;
+    } catch (error: unknown) {
+      console.warn(`Error reading sessionStorage key “${key}”:`, error);
+      return initialValueToUse;
+    }
+  }, [initialValue, key, deserializer]);
+
+  const [storedValue, setStoredValue] = useState(() => {
+    if (initializeWithValue) {
+      return readValue();
+    }
+
+    return initialValue instanceof Function ? initialValue() : initialValue;
+  });
+
+  // Return a wrapped version of useState's setter function that ...
+  // ... persists the new value to sessionStorage.
+  const setValue: Dispatch<SetStateAction<T>> = useCallback(
+    (value: SetStateAction<T>) => {
+      try {
+        // Allow value to be a function so we have the same API as useState
+        const newValue = value instanceof Function ? value(readValue()) : value;
+
+        // Save to session storage
+        window.sessionStorage.setItem(key, serializer(newValue));
+
+        // Save state
+        setStoredValue(newValue);
+
+        // We dispatch a custom event so every similar useSessionStorage hook is notified
+        window.dispatchEvent(
+          new StorageEvent("authgear-portal:session-storage", { key })
+        );
+      } catch (error: unknown) {
+        console.warn(`Error setting sessionStorage key “${key}”:`, error);
+      }
+    },
+    [key, readValue, serializer]
+  );
+
+  const removeValue = useCallback(() => {
+    const defaultValue =
+      initialValue instanceof Function ? initialValue() : initialValue;
+
+    // Remove the key from session storage
+    window.sessionStorage.removeItem(key);
+
+    // Save state with default value
+    setStoredValue(defaultValue);
+
+    // We dispatch a custom event so every similar useSessionStorage hook is notified
+    window.dispatchEvent(
+      new StorageEvent("authgear-portal:session-storage", { key })
+    );
+  }, [initialValue, key]);
+
+  useEffect(() => {
+    setStoredValue(readValue());
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [key]);
+
+  const handleStorageChange = useCallback(
+    (event: StorageEvent | CustomEvent) => {
+      if ((event as StorageEvent).key && (event as StorageEvent).key !== key) {
+        return;
+      }
+      setStoredValue(readValue());
+    },
+    [key, readValue]
+  );
+
+  // this only works for other documents, not the current one
+  useEventListener("storage", handleStorageChange);
+
+  // this is a custom event, triggered in writeValueToSessionStorage
+  // See: useSessionStorage()
+  useEventListener("authgear-portal:session-storage", handleStorageChange);
+
```

Comment:
It seems the custom event is for listening to changes in the storage and reflect the changes immediately in the screen. But do we need this? We only want to read the storage once when we enter  the screen?

Reply:
I had the same thought as you, but I kept these parts because I think this logic might serves its purpose if one day we have 2 components sharing 1 shared session storage key.

Still, lemme remove it. 🙏